File: Python/48.Rotate Image.py
#将二维矩阵顺时针旋转90度

#不开辟空间的做法
class Solution(object):
    def rotate(self, matrix):
        L = len(matrix)
        R = (L + 1) // 2
        for x in range(0, R):
            for y in range(0, L - R):
                #(x,y)->(y,l-1-x)->(l-1-x,l-1-y)->(l-1-y,x)
                matrix[x][y], matrix[y][L-1-x], matrix[L-1-x][L-1-y], matrix[L-1-y][x] \
                = matrix[L-1-y][x], matrix[x][y], matrix[y][L-1-x], matrix[L-1-x][L-1-y]
        return matrix

#重新开辟空间的做法
import copy
class Solution1(object):
    def rotate(self, matrix):
        n = len(matrix)
        matrixx = copy.deepcopy(matrix)
        for i in range(n):
            for j in range(n):
                matrixx[j][n-i-1] = matrix[i][j]
        return matrixx

s = Solution()
print(s.rotate([[1,2,3,4],[5,6,7,8],[9,10,11,12],[13,14,15,16]]))
print(s.rotate([[1,2,3],[4,5,6],[7,8,9]]))
print(s.rotate([[1,2],[3,4]]))

File: C++/Array/48.Rotate Image.cpp
/*
 n x n二维数组，顺时针旋转90度。
*/


#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	void rotate(vector<vector<int>>& matrix) {
		int n = matrix.size();
        for(int level=0; level < n/2; ++level)
        {
        	int m = n-1;
        	for(int i = level; i < m-level; ++i)
        	{
        		int temp = matrix[level][i];
        		matrix[level][i] = matrix[m-i][level];
        		matrix[m-i][level] = matrix[m-level][m-i];
        		matrix[m-level][m-i] = matrix[i][m-level];
        		matrix[i][m-level] = temp;
        	}
        }
    }
};

int main()
{
	Solution s;
	vector<vector<int>> nums{{1,2,3,4},{5,6,7,8},{9,10,11,12},{13,14,15,16}};
	s.rotate(nums);
	for(int i=0 ; i < nums.size(); ++i)
	{
		for(int j=0; j < nums[i].size(); ++j)
		{
			cout<<nums[i][j]<<"\t";
		}
		cout<<endl;
	}
}

